backend/instancemgmt/locker.go

```diff
@@ -0,0 +1,70 @@
+package instancemgmt
+
+import (
+	"fmt"
+	"sync"
+)
+
+type locker struct {
+	locks   map[interface{}]*sync.RWMutex
+	locksRW *sync.RWMutex
+}
+
+func newLocker() *locker {
+	return &locker{
+		locks:   make(map[interface{}]*sync.RWMutex),
+		locksRW: new(sync.RWMutex),
+	}
+}
+
+func (lkr *locker) Lock(key interface{}) {
+	lk, ok := lkr.getLock(key)
+	if !ok {
+		lk = lkr.newLock(key)
+	}
+	lk.Lock()
+}
+
+func (lkr *locker) Unlock(key interface{}) {
+	lk, ok := lkr.getLock(key)
+	if !ok {
+		panic(fmt.Errorf("lock for key '%s' not initialized", key))
+	}
+	lk.Unlock()
+}
+
+func (lkr *locker) RLock(key interface{}) {
+	lk, ok := lkr.getLock(key)
+	if !ok {
+		lk = lkr.newLock(key)
+	}
+	lk.RLock()
+}
+
+func (lkr *locker) RUnlock(key interface{}) {
+	lk, ok := lkr.getLock(key)
+	if !ok {
+		panic(fmt.Errorf("lock for key '%s' not initialized", key))
```

Comment:
It is implemented to behave the same way golang RWMutex behaves:
https://golang.org/pkg/sync/#RWMutex.RUnlock
`It is a run-time error if rw is not locked for reading on entry to RUnlock.`

Reply:
Fair point. Comment on the methods would be nice explaining what they do for better understanding of this possible panic-behavior since they're public and not internal to the package.

Reply:
Provided comments.